### PR TITLE
KAFKA-6774: Improve the default group id behavior in KafkaConsumer (KIP-289)

### PR DESCRIPTION
**Pull request by vahidhashemian**
Improve the default group id behavior by
* changing the default consumer group to null, where no offset commit or fetch, or group management operations are allowed
* deprecate the use of empty (`""`) consumer group on the client

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 1]
```diff
@@ -2216,6 +2231,11 @@ private void throwIfNoAssignorsConfigured() {
                 ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG + " configuration property");
     }
 
+    private void maybeThrowInvalidGroupIdException() {
+        if (groupId == null)
+            throw new InvalidGroupIdException("The null group id does not support group management operations.");
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Perhaps group management or the the offset commit api?

**Review by hachikuji — COMMENTED** [review]
@vahidhashemian Thanks for the PR. I think we also might need to hook into the logic for setting the initial position of the consumer. By default, we attempt to fetch committed offsets, but if the group id is null, we should just skip this step.

**Comment by vahidhashemian** [comment]
@hachikuji Thanks for reviewing the PR. Submitted a patch to address your feedback.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -662,10 +664,15 @@ private KafkaConsumer(ConsumerConfig config,
...
             this.clientId = clientId;
-            String groupId = config.getString(ConsumerConfig.GROUP_ID_CONFIG);
-
+            this.groupId = config.getString(ConsumerConfig.GROUP_ID_CONFIG);
             LogContext logContext = new LogContext("[Consumer clientId=" + clientId + ", groupId=" + groupId + "] ");
             this.log = logContext.logger(getClass());
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            if (groupId == null)
+                // overwrite in case of default group id
+                enableAutoCommit = false;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: I think it's fine to change the default as long as the user hasn't explicitly provided the configuration. However, if they have explicitly set `enable.auto.commit=true`, then it might be better to raise an exception.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 3]
```diff
@@ -1271,6 +1281,57 @@ public void closeShouldBeIdempotent() {
         consumer.close();
     }
 
+    @Test
+    public void testOperationsBySubscribingConsumerWithDefaultGroupId() {
+        try {
+            newConsumer((String) null).subscribe(Collections.singleton(topic));
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1297]: We should probably have a `fail` below if this returns without errors?

**Review by hachikuji — COMMENTED** [review]
Left a couple more comments. Do we have an integration test that covers basic fetching using a null group id? We should probably make sure we have at least one case which uses null and one which uses the empty string.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 3 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:1297]: Thanks! Totally missed them.

**Review by vahidhashemian — COMMENTED** [review]
@hachikuji I tried to address your recent comments in the new commit. Thanks for the additional feedback.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 4]
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+    producer.close()
+
+    // consumer 1 uses the default group id and consumes from earliest offset
+    val consumer1Config = new Properties(consumerConfig)
+    consumer1Config.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
+    consumer1Config.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+    consumer1Config.put(ConsumerConfig.CLIENT_ID_CONFIG, "consumer1")
+    val consumer1 = createConsumer(
+      configOverrides = consumer1Config,
+      configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:1841]: Is it possible to just pass "groupId=null" as an override?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 5]
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+
+    val producer = createProducer()
+    producer.send(new ProducerRecord(topic, partition, "k1".getBytes, "v1".getBytes)).get()
+    producer.send(new ProducerRecord(topic, partition, "k2".getBytes, "v2".getBytes)).get()
+    producer.send(new ProducerRecord(topic, partition, "k3".getBytes, "v3".getBytes)).get()
+    producer.close()
+
+    // consumer 1 uses the default group id and consumes from earliest offset
+    val consumer1Config = new Properties(consumerConfig)
+    consumer1Config.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I suppose we could leave this out?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 6]
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+    // consumer 2 uses the default group id and consumes from latest offset
+    val consumer2Config = new Properties(consumerConfig)
+    consumer2Config.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
+    consumer2Config.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
+    consumer2Config.put(ConsumerConfig.CLIENT_ID_CONFIG, "consumer2")
+    val consumer2 = createConsumer(
+      configOverrides = consumer2Config,
+      configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
+
+    // consumer 3 uses the default group id and consumes from earliest offset
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Seems more accurate to say that consumer 3 begins from an explicit seek?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 7]
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+
+    val records3 = consumer3.poll(Duration.ofMillis(5000))
+    consumer3.commitSync()
+
+    consumer1.close()
+    consumer2.close()
+    consumer3.close()
+
+    assertTrue("Expected consumers to consumer a single message each from the last committed offset",
+      records1.count() == 1 && records1.records(tp).asScala.head.offset == 0 &&
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: nit: maybe better to turn this into explicit equals assertions so that we know which check failed.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 8]
```diff
@@ -1813,4 +1814,128 @@ class PlaintextConsumerTest extends BaseConsumerTest {
...
+    consumer2.assign(asList(tp))
+    consumer3.assign(asList(tp))
+
+    val records1 = consumer1.poll(Duration.ofMillis(5000))
+    consumer1.commitSync()
+
+    val records2 = consumer2.poll(Duration.ofMillis(5000))
+    consumer2.commitSync()
+
+    val records3 = consumer3.poll(Duration.ofMillis(5000))
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: The third iteration here feels redundant. Doesn't it just read from the committed offset like `consumer2`?

**Review by hachikuji — COMMENTED** [review]
Thanks for the tests. Left a few minor comments.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 5 above)
Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: That's correct.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 4 above)
Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:1841]: Doesn't seem so, I get a NPE:
```
java.lang.NullPointerException
	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1021)
	at java.base/java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1016)
	at java.base/java.util.Properties.put(Properties.java:1309)
	at kafka.api.PlaintextConsumerTest.testConsumingWithNullGroupId(PlaintextConsumerTest.scala:1838)
        ...
```

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 8 above)
Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Yes, I put it to verify the `auto.offset.reset` doesn't affect the behavior. I'll remove it in the new commit.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 7 above)
Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Sounds good.

**Review by vahidhashemian — COMMENTED** [review]
@hachikuji I updated the unit tests based on your comments. Thanks!

**Review by hachikuji — COMMENTED** [review]
@vahidhashemian Thanks for the patch. Looks good overall. I had just one thought. When the groupId is null, there should be no coordinator interaction at all. We could make this explicit by not bothering to construct the `ConsumerCoordinator` object. That would take it out of the poll path and reduce the risk of using it accidentally. What do you think?

**Comment by vahidhashemian** [comment]
@hachikuji thanks for the feedback. Your suggestion makes sense. Even though it's not explicitly mentioned in the KIP, I think it wouldn't result in a different behavior that what is covered by the KIP. I submitted a patch that should cover it (it did not break any test locally), but please let me know if I missed something. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 9]
```diff
@@ -491,6 +491,9 @@ public boolean rejoinNeededOrPending() {
      * @return true iff the operation completed within the timeout
      */
     public boolean refreshCommittedOffsetsIfNeeded(Timer timer) {
+        // no committed offset should be fetched for the default group id
+        if (groupId == null) return true;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Perhaps now we don't need this? We could add an assertion in the constructor that the groupId is not null.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 10]
```diff
@@ -203,7 +206,7 @@ public void testInvalidSocketReceiveBufferSize() {
 
     @Test
     public void testSubscription() {
-        KafkaConsumer<byte[], byte[]> consumer = newConsumer();
+        KafkaConsumer<byte[], byte[]> consumer = newConsumer("");
```
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: nit: not sure if there's a strong reason for any of these to use the empty group id. Could we use `groupId` instead?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 11]
```diff
@@ -2216,6 +2228,11 @@ private void throwIfNoAssignorsConfigured() {
                 ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG + " configuration property");
     }
 
+    private void maybeThrowInvalidGroupIdException() {
+        if (groupId == null)
+            throw new InvalidGroupIdException("The null group id does not support group management or offset commit APIs.");
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: I was wondering if we could rephrase this to suggest what the user should do more clearly. For example, "To use the group management or offset commit APIs, you must provide a valid `group.id` in the consumer configuration."

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala [hunk 12]
```diff
@@ -1814,4 +1815,110 @@ class PlaintextConsumerTest extends BaseConsumerTest {
         s"The current assignment is ${consumer.assignment()}")
   }
 
+  @Test
+  def testConsumingWithNullGroupId(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:1819]: Just to complete the integration picture, could we make a call to `committed` and assert the exception?

**Review by hachikuji — COMMENTED** [review]
Thanks for the updates, just a few additional comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 9 above)
Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: You're right. I added an assertion in `AbstractCoordinator` constructor. It seemed to be a better fit there.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 11 above)
Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Sounds good, thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java (hunk 10 above)
Comment (vahidhashemian) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: No particular reason other than I wanted to stick to the actual group id that was used before. I updated these to use the provided `groupId`.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala (hunk 12 above)
Comment (vahidhashemian) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:1819]: I did that in the next commit.

**Review by vahidhashemian — COMMENTED** [review]
@hachikuji thanks for the feedback/suggestions. The new commit should address them.

**Comment by hachikuji** [comment]
@vahidhashemian Can you look into the build failures? I looked at one of the streams tests (`KTableSourceTopicRestartIntegrationTest`)  and saw there was some dependence on the empty group id.
```
Exception in thread "ktable-restore-from-source-2639ce74-b612-423d-aaff-96cc59653aa7-StreamThread-1" org.apache.kafka.streams.errors.StreamsException: stream-thread [ktable-restore-from-source-2639ce74-b612-423d-aaff-96cc59653aa7-StreamThread-1] Failed to rebalance.
	at org.apache.kafka.streams.processor.internals.StreamThread.pollRequests(StreamThread.java:922)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:822)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:777)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:747)
Caused by: org.apache.kafka.common.errors.InvalidGroupIdException: To use the group management or offset commit APIs, you must provide a valid group.id in the consumer configuration.
```
cc @guozhangwang

**Comment by vahidhashemian** [comment]
Thanks @hachikuji for catching this. 
I see this code in `StreamsConfig.getRestoreConsumerConfigs`:
```
        // no need to set group id for a restore consumer
        baseConsumerProps.remove(ConsumerConfig.GROUP_ID_CONFIG);
```
which now means setting the group id to null. I think it should be safe to replace it with
```
        // no need to use a specific group id for a restore consumer
        baseConsumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "");
```
Thoughts?
@guozhangwang

**Comment by hachikuji** [comment]
The exception is being raised from a call to `unsubscribe`. We currently allow this API to be used in order to remove all assigned partitions, so I think we should probably remove the group id validation from that API.

**Comment by vahidhashemian** [comment]
Sounds good, and seems to have no side effect with respect to existing unit tests. My initial thought was to use the empty string to keep the tests run with the same config as before. I'll submit a patch.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 13]
```diff
@@ -654,19 +656,23 @@ public KafkaConsumer(Properties properties,
...
             this.clientId = clientId;
-            String groupId = config.getString(ConsumerConfig.GROUP_ID_CONFIG);
-
+            this.groupId = config.getString(ConsumerConfig.GROUP_ID_CONFIG);
             LogContext logContext = new LogContext("[Consumer clientId=" + clientId + ", groupId=" + groupId + "] ");
             this.log = logContext.logger(getClass());
 
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            if (groupId == null) { // overwrite in case of default group id where the config is not explicitly provided
+                if (!config.originals().containsKey(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG))
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:671]: Shouldn't we handle the `else` case? I was thinking that if the user enables auto commit explicitly, but provides no group id, then we should raise a configuration exception.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 14]
```diff
@@ -136,6 +136,7 @@ public AbstractCoordinator(LogContext logContext,
                                Time time,
                                long retryBackoffMs,
                                boolean leaveGroupOnClose) {
+        assert groupId != null : "Expected a non-null group id for coordinator construction";
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: nit: can we use `Objects.requireNonNull` instead?

**Review by hachikuji — COMMENTED** [review]
Thanks for the updates. Looks like the build is passing. Just two more comments and I think we can merge.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 13 above)
Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:671]: What I had here was to let the consumer run and throw an exception when a commit is triggered (or even earlier when e.g. `subscribe` is called). But I think your suggestion makes more sense because it points directly to the root cause, i.e. the misconfiguration. Updated in the new patch.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 14 above)
Comment (vahidhashemian) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Made this change too.

**Review by vahidhashemian — COMMENTED** [review]
@hachikuji Thanks for the feedback. Responded to your comment inline and submitted a new patch.

**Review by hachikuji — APPROVED** [review]
LGTM. Pushed a trivial fix for the null check in `AbstractCoordinator`.
